### PR TITLE
Migrate erc20-contracts to foundry

### PR DESCRIPTION
**Pull request by elenadimitrova**
## Description
Migrate the protocol/contracts/erc20 to foundry for compilation and testing. Removes the obsoleted typescript / node/ custom tooling setup and also rips out any unnecessary contracts. 

Note that the interfaces that are compatible between solidity versions have been merged, that is
`IERC20TokenV06` and `IERC20TokenV08` -> `IERC20Token`
 `IEtherTokenV06` and `IEtherTokenV08` -> `IEtherToken`

Removed contracts have been used only in exchange-v3 which is being obsoleted 
contracts/erc20/contracts/src/LibERC20Token.sol
contracts/erc20/contracts/src/interfaces/IEtherToken.sol (used in v3 staking)
contracts/erc20/contracts/test/DummyMultipleReturnERC20Token.sol
contracts/erc20/contracts/test/DummyNoReturnERC20Token.sol
contracts/erc20/contracts/test/TestLibERC20Token.sol
contracts/erc20/contracts/test/TestLibERC20TokenTarget.sol
contracts/erc20/contracts/test/UntransferrableDummyERC20Token.sol

We have also removed the `DummyERC20Token` as this can be replaced by either using the `ZRXToken` contract where necessary or just the generic open-zeppelin ERC20 implementation. Note that this also [implements infinite approvals](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/ERC20.sol#L326) so there should be no difference in functionally. Additionally the `ZRXToken` is also an infinite allowance token itself.

Closes https://linear.app/0xproject/issue/PRO-44/zrx-artifact-is-incompatible-with-foundry

### CODE REVIEW

File: contracts/erc20/test/ZRXTokenTest.t.sol [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+pragma solidity ^0.8.17;
+
+import "forge-std/Test.sol";
+import "../src/IERC20Token.sol";
+
+contract ZRXTokenTest is Test {
+    address payable internal owner = payable(vm.addr(1));
+    address payable internal user = payable(vm.addr(2));
+    uint256 internal totalSupply = 1_000_000_000 * 1e18;
+    uint256 internal constant MAX_UINT = 2 ^ (256 - 1);
```
Comment (duncancmt) [inline contracts/erc20/test/ZRXTokenTest.t.sol:24]: use `type(uint256).max`

File: contracts/erc20/test/ZRXTokenTest.t.sol [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    IERC20Token zrxToken;
+
+    function setUp() public {
+        vm.deal(owner, 1e20);
+        vm.deal(user, 1e20);
+
+        vm.prank(owner);
+        bytes memory _bytecode = abi.encodePacked(vm.getCode("./out/ZRXToken.sol/ZRXToken.json"));
```
Comment (duncancmt) [inline contracts/erc20/test/ZRXTokenTest.t.sol:32]: `vm.getCode` returns `bytes memory`. Why is `abi.encodePacked` used?

File: contracts/erc20/test/ZRXTokenTest.t.sol [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+        assertEq(zrxToken.balanceOf(user), 100);
+        assertEq(zrxToken.balanceOf(owner), totalSupply - 100);
+    }
+
+    function testShouldReturnTrueOnAZeroValueTransfer() public {
+        vm.prank(owner);
+        bool success = zrxToken.transfer(user, 0);
+        assertEq(success, true);
+    }
```
Comment (duncancmt) [inline contracts/erc20/test/ZRXTokenTest.t.sol:65]: Consider also testing that this works when called by a user with zero balance

File: contracts/erc20/test/ZRXTokenTest.t.sol [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        zrxToken.transfer(user, 100);
+
+        assertEq(zrxToken.balanceOf(user), 100);
+        assertEq(zrxToken.balanceOf(owner), totalSupply - 100);
+    }
+
+    function testShouldReturnTrueOnAZeroValueTransfer() public {
+        vm.prank(owner);
+        bool success = zrxToken.transfer(user, 0);
+        assertEq(success, true);
```
Comment (duncancmt) [inline contracts/erc20/test/ZRXTokenTest.t.sol:64]:
```suggestion
        assertTrue(success);
```

**Review by duncancmt — APPROVED** [review]
